File: parsley/jvm/src/main/java/parsley/internal/machine/instructions/JitImpl.java
/*
 * Copyright 2020 Parsley Contributors <https://github.com/j-mie6/Parsley/graphs/contributors>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */
package parsley.internal.machine.instructions;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface JitImpl {
    boolean noop() default false;
    
    int consumeOperands() default 0;

    IntKind intReturnKind() default IntKind.Pc;
    
    Action[] beforeActions() default {};
    
    Action[] afterActions() default {};

    String[] constants() default {};

    Param[] params() default {};

    int[] updateCheckOffsets() default {};
    
    enum Action {
        PushTrue,
        Swap,
        DupX1,
        Dup
    }

    enum Param {
        Pc,
        HandlerCheck
    }

    enum IntKind {
        Pc,
        Char,
        CodePoint
    }
}

File: parsley/js-native/src/main/scala/parsley/internal/machine/instructions/JitImpl.scala
package parsley.internal.machine.instructions

import scala.annotation.StaticAnnotation

final class JitImpl(val noop: Boolean = false,
                    val consumeOperands: Int = 0,
                    val intReturnKind: JitImpl.IntKind.Value = JitImpl.IntKind.Pc,
                    val beforeActions: Array[JitImpl.Action.Value] = Array(),
                    val afterActions: Array[JitImpl.Action.Value] = Array(),
                    val constants: Array[String] = Array(),
                    val params: Array[JitImpl.Param.Value] = Array(),
                    val updateCheckOffsets: Array[Int] = Array()
                   ) extends StaticAnnotation

object JitImpl {
    object Action extends Enumeration {
        val PushTrue, Swap, DupX1, Dup = Value
    }

    object Param extends Enumeration {
        val Pc, HandlerCheck = Value
    }

    object IntKind extends Enumeration {
        val Pc, Char, CodePoint = Value
    }
}
